File: Data Structures & Algorithms/combination-target-sum/submission-5.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res=[]
        n = len(nums)
        nums.sort()
        def backtracking(i:int,curr,total):
            if total==target:
                res.append(curr.copy())
                return
         
            for j in range(i, len(nums)):
                if total + nums[j] > target:
                    return
                curr.append(nums[j])
                backtracking(j, curr, total + nums[j])
                curr.pop()
            
        backtracking(0,[],0)
        return res
            

File: Data Structures & Algorithms/subsets/submission-5.cpp
class Solution {
    void dfs(vector<int>& nums, int i, vector<int> curr,vector<vector<int>>& res) {
        if (i >= nums.size()) {
            res.push_back(curr);
            return;
        }
        curr.push_back(nums[i]);

        dfs(nums, i + 1, curr, res);
        curr.pop_back();
        dfs(nums, i + 1, curr, res);
    }

   public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> subset;
        dfs(nums, 0, subset, res);
        return res;
    }
};
